### PR TITLE
Add support for huaweicloud dns

### PR DESCRIPTION
**Pull request by huihuimoe**
Hi guys, I'm trying out dnscontrol with huaweicloud dns, but dnscontrol doesn't support it, so I wrote it.

This is the homepage of huaweicloud dns, please check it out
https://www.huaweicloud.com/intl/en-us/product/dns.html

I ran the built-in integrationTest test and it should all pass.
https://paste.ubuntu.com/p/2Wv8rcJJs4/

There is still some work left to do, I will add it later.
- [x] write document

If you have any questions, please feel free to ask, thanks!

### CODE REVIEW

File: providers/huaweicloud/convert.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
+package huaweicloud
+
+// Convert the provider's native record description to models.RecordConfig.
```
Comment (tlimoncelli) [inline providers/huaweicloud/convert.go:3]:
```suggestion
// Support for Huawei Cloud DNS.
// API Documentation: https://www.huaweicloud.com/intl/en-us/product/dns.html
```

File: providers/huaweicloud/records.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+		return "", err
+	}
+	marker := url.Query().Get("marker")
+	if marker == "" {
+		return "", fmt.Errorf("marker not found in URL %s", link)
+	}
+	return marker, nil
+}
+
+func (c *huaweicloudProvider) getZoneRecords(zoneID string) (*[]model.ListRecordSets, error) {
```
Comment (tlimoncelli) [inline providers/huaweicloud/records.go:184]: Please rename this function.  having GetZoneRecords() and getZoneRecords(), while valid in Go, could be confusing.

Thanks!

**Comment by tlimoncelli** [comment]
Hello!

Thank you for the contribution!  It looks excellent!

I made a few minor suggestions.  Once the documentation is finalized it should be ready for merge.

Thanks again!

**Comment by huihuimoe** [comment]
Hi,
I've added documentation, as well as modified the suggestions.

File: .github/workflows/pr_test.yml [hunk 3]
```diff
@@ -88,7 +88,7 @@ jobs:
         Write-Host "Integration test providers: $Providers"
         echo "integration_test_providers=$(ConvertTo-Json -InputObject $Providers -Compress)" >> $env:GITHUB_OUTPUT
       env:
-        PROVIDERS: "['AZURE_DNS','BIND','BUNNY_DNS','CLOUDFLAREAPI','CLOUDNS','DIGITALOCEAN','GANDI_V5','GCLOUD','HEDNS','HEXONET','INWX','NAMEDOTCOM','NS1','POWERDNS','ROUTE53','TRANSIP']"
+        PROVIDERS: "['AZURE_DNS','BIND','BUNNY_DNS','CLOUDFLAREAPI','CLOUDNS','DIGITALOCEAN','GANDI_V5','GCLOUD','HEDNS','HEXONET','INWX','NAMEDOTCOM','NS1','POWERDNS','ROUTE53','TRANSIP','HUAWEICLOUD']"
```
Comment (cafferata) [inline .github/workflows/pr_test.yml:91]: Would you please apply these in alphabetical order?

File: documentation/provider/huaweicloud.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+If you don't know how to fill in the `Region` parameter, you can refer to the [endpoint page of huaweicloud](https://developer.huaweicloud.com/intl/en-us/endpoint?DNS). For reference, I'm using on its international site, the `Region` I need to fill in is `ap-southeast-1`.
+
+If you're still not sure, you can log into Huaweicloud's website and open the [API Explorer](https://console-intl.huaweicloud.com/apiexplorer/#/openapi/DNS/debug?api=ListPublicZones), find the `ListPublicZones` API, select a different Region and click Debug to try and find your Region!
+
+## New domains
+If a domain does not exist in your Huawei Cloud account, DNSControl will automatically add it with the `push` command. Don't worry about incurring fees, it's free!
+
+## GeoDNS
+Managing GeoDNS RRSet on huawei cloud (also called **Line** in Huawei Cloud DNS) is not supported in dnscontrol.
```
Comment (cafferata) [inline documentation/provider/huaweicloud.md:72]: We write `dnscontrol` as follows: DNSControl.

**Review by cafferata — CHANGES_REQUESTED** [review]

File: .github/workflows/pr_test.yml (hunk 3 above)
Comment (huihuimoe) [reply .github/workflows/pr_test.yml:91]: Sure

Comment (cafferata) [reply .github/workflows/pr_test.yml]: Thanks!

File: documentation/provider/huaweicloud.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+D("example.com", REG_NONE, DnsProvider(DSP_HWCLOUD),
+    A("test", "1.2.3.4"),
+END);
+```
+{% endcode %}
+
+## Activation
+DNSControl depends on a standard [IAM User](https://support.huaweicloud.com/intl/en-us/usermanual-iam/iam_02_0003.html) with permission to list, create and update hosted zones.
+
+You can apply the `DNS FullAccess` policy however this includes access to many other areas. The minimum permissions required are as follows:
```
Comment (tlimoncelli) [inline documentation/provider/huaweicloud.md:42]:
```suggestion
The `DNS FullAccess` policy will also work, but that provides access to many other areas and violates the "principle of least privilege".

The minimum permissions required are as follows:
```

File: documentation/provider/huaweicloud.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                "dns:zone:list",
+                "dns:recordset:update",
+                "dns:recordset:list"
+            ]
+        }
+    ]
+}
+```
+
+If you don't know how to fill in the `Region` parameter, you can refer to the [endpoint page of huaweicloud](https://developer.huaweicloud.com/intl/en-us/endpoint?DNS). For reference, I'm using on its international site, the `Region` I need to fill in is `ap-southeast-1`.
```
Comment (tlimoncelli) [inline documentation/provider/huaweicloud.md:64]:
```suggestion
To determine the `Region` parameter, refer to the [endpoint page of huaweicloud](https://developer.huaweicloud.com/intl/en-us/endpoint?DNS). For example, on the international site, the `Region` name `ap-southeast-1` is known to work.
```

File: documentation/provider/huaweicloud.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+If you're still not sure, you can log into Huaweicloud's website and open the [API Explorer](https://console-intl.huaweicloud.com/apiexplorer/#/openapi/DNS/debug?api=ListPublicZones), find the `ListPublicZones` API, select a different Region and click Debug to try and find your Region!
```
Comment (tlimoncelli) [inline documentation/provider/huaweicloud.md:66]:
```suggestion
If that doesn't work, log into Huaweicloud's website and open the [API Explorer](https://console-intl.huaweicloud.com/apiexplorer/#/openapi/DNS/debug?api=ListPublicZones), find the `ListPublicZones` API, select a different Region and click Debug to try and find your Region.
```

**Comment by huihuimoe** [comment]
Works perfectly on integrationTest and on my own domain now.
it's ready for merge, any other suggestions?

**Review by tlimoncelli — APPROVED** [review]

**Comment by tlimoncelli** [comment]
Looks good!  I'll merge soon.

If you would like HUAWEICLOUD to be automatically tested as part of each release, please follow the instructions here: https://docs.dnscontrol.org/developer-info/byo-secrets 

Thanks for contributing this!  I'm sure a lot of people will find it useful!

**Pull request merged**